File: src/sentinel/api/routers/schedule.py
from __future__ import annotations

from datetime import date

from fastapi import APIRouter, Depends, Query

from sentinel.api.deps import get_decision_scope, get_settings, make_page_params_dependency
from sentinel.api.schemas.common import DecisionScope, Page
from sentinel.api.schemas.scheduling import BacklogRowOut, ReplanningRunOut, ScheduleRowOut
from sentinel.api.services import scheduling_service
from sentinel.api.services.pagination import PageParams
from sentinel.config import Settings

router = APIRouter(prefix="/v1/schedule", tags=["schedule"])

_schedule_page = make_page_params_dependency(default_sort="final_policy_rank")
_backlog_page = make_page_params_dependency(default_sort="backlog_position")


@router.get("", response_model=Page[ScheduleRowOut])
def list_schedule(
    establishment_id: str | None = Query(None),
    schedule_status: str | None = Query(None),
    scheduled_date: date | None = Query(None),
    scope: DecisionScope = Depends(get_decision_scope),
    page: PageParams = Depends(_schedule_page),
    settings: Settings = Depends(get_settings),
) -> Page[ScheduleRowOut]:
    return scheduling_service.get_schedule(
        settings,
        scope,
        page,
        establishment_id=establishment_id,
        schedule_status=schedule_status,
        scheduled_date=scheduled_date,
    )


@router.get("/dates", response_model=list[dict[str, object]])
def list_schedule_dates(
    scope: DecisionScope = Depends(get_decision_scope), settings: Settings = Depends(get_settings)
) -> list[dict[str, object]]:
    return scheduling_service.get_schedule_dates(settings, scope)


@router.get("/backlog", response_model=Page[BacklogRowOut])
def list_backlog(
    scope: DecisionScope = Depends(get_decision_scope),
    page: PageParams = Depends(_backlog_page),
    settings: Settings = Depends(get_settings),
) -> Page[BacklogRowOut]:
    return scheduling_service.get_backlog(settings, scope, page)


@router.get("/summary", response_model=list[dict[str, object]])
def get_summary(
    scope: DecisionScope = Depends(get_decision_scope), settings: Settings = Depends(get_settings)
) -> list[dict[str, object]]:
    return scheduling_service.get_small_table(settings, scope, table="schedule_summary")


@router.get("/capacity-utilization", response_model=list[dict[str, object]])
def get_capacity_utilization(
    scope: DecisionScope = Depends(get_decision_scope), settings: Settings = Depends(get_settings)
) -> list[dict[str, object]]:
    return scheduling_service.get_small_table(settings, scope, table="capacity_utilization")


@router.get("/priority-preservation", response_model=list[dict[str, object]])
def get_priority_preservation(
    scope: DecisionScope = Depends(get_decision_scope), settings: Settings = Depends(get_settings)
) -> list[dict[str, object]]:
    return scheduling_service.get_small_table(settings, scope, table="priority_preservation")


@router.get("/replanning-runs", response_model=list[ReplanningRunOut])
def get_replanning_runs(
    scope: DecisionScope = Depends(get_decision_scope), settings: Settings = Depends(get_settings)
) -> list[ReplanningRunOut]:
    return scheduling_service.get_replanning_runs(settings, scope)


__all__ = ["router"]

File: frontend/src/api/schedule.ts
import { apiFetch } from './client'
import type { PageParams } from './recommendations'
import type {
  BacklogRowOut,
  DecisionScope,
  Page,
  ReplanningRunOut,
  ScheduleRowOut,
} from './types'

export interface ScheduleFilters {
  establishment_id?: string
  schedule_status?: string
  scheduled_date?: string
}

export interface ScheduleDateOut {
  scheduled_date: string
  n_establishments: number
}

export function listSchedule(
  scope: DecisionScope,
  filters: ScheduleFilters,
  page: PageParams,
  signal?: AbortSignal,
): Promise<Page<ScheduleRowOut>> {
  return apiFetch('/v1/schedule', { ...scope, ...filters, ...page }, signal)
}

/** Every date this plan actually has inspections on, each with a count -- the real,
 * establishment-backed set a day picker should offer, never a generic calendar the user could
 * pick an empty day from. */
export function listScheduleDates(
  scope: DecisionScope,
  signal?: AbortSignal,
): Promise<ScheduleDateOut[]> {
  return apiFetch('/v1/schedule/dates', { ...scope }, signal)
}

export function listBacklog(
  scope: DecisionScope,
  page: PageParams,
  signal?: AbortSignal,
): Promise<Page<BacklogRowOut>> {
  return apiFetch('/v1/schedule/backlog', { ...scope, ...page }, signal)
}

export function getScheduleSummary(
  scope: DecisionScope,
  signal?: AbortSignal,
): Promise<Record<string, unknown>[]> {
  return apiFetch('/v1/schedule/summary', { ...scope }, signal)
}

export function getCapacityUtilization(
  scope: DecisionScope,
  signal?: AbortSignal,
): Promise<Record<string, unknown>[]> {
  return apiFetch('/v1/schedule/capacity-utilization', { ...scope }, signal)
}

export function getPriorityPreservation(
  scope: DecisionScope,
  signal?: AbortSignal,
): Promise<Record<string, unknown>[]> {
  return apiFetch('/v1/schedule/priority-preservation', { ...scope }, signal)
}

export function listReplanningRuns(
  scope: DecisionScope,
  signal?: AbortSignal,
): Promise<ReplanningRunOut[]> {
  return apiFetch('/v1/schedule/replanning-runs', { ...scope }, signal)
}
